plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+
+-- | Heap location. Int should always be big enough: maxBound::Int is 2^63-1
+type HeapLoc = Int
+
+-- | Heap entries: unevaluated and evaluated lazy function arguments.
+data HeapEntry =
+    Evaluated   Closure  -- The term in this closure should always be a value. It would be good if the type system could distiniguish these.
+  | Unevaluated Closure
+
+-- | The heap itself: a map from heap locations to
+-- unevaluated/evaluated terms, together with an integer counter
+-- @_top@ which serves as the current heap location.
+data Heap = Heap {
+      _heap :: IntMap HeapEntry
+    , _top  :: Int
+}
+
+{- Environments map variable ids to locations, the heap maps locations
+   to terms.  We really do require this indirection: there may be
+   multiple objects in the heap corresponding to the same variable
+   (during recursion, for example).  For the same reason we always
+   work with closures instead of terms; this also makes it easier
+   to ensure that we're using the correct environment. -}
+
+
+-- | Stack frames.  These are effectively continuations: would using explicit CPS speed things up?
+-- The spec also has a frame for n-ary built in application, but that's not in the AST yet
+data Frame
+    = FrameAppArg Closure                        -- ^ @[_ N]@       -- Tells the machine that we've got back to N after evaluating M in [M N]
+    | FrameHeapUpdate HeapLoc                    -- ^ @(update l)@  -- Mark the stack for evaluation of a delayed argument
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+      deriving (Show)
+
+-- | A context is a stack of frames.
+type EvaluationContext = [Frame]
+
+-- | A local version of the EvaluationResult type using closures instead of values
+data LMachineResult
+    = Success Closure Heap  -- We need both the environment and the heap to see what the value "really" is.
+    | Failure
+
+emptyEnvironment :: Environment
+emptyEnvironment = Environment IntMap.empty
+
+-- | Extend an environment with a new binding
+updateEnvironment :: Int -> HeapLoc -> Environment -> Environment
+updateEnvironment index cl (Environment m) = Environment (IntMap.insert index cl m)
+
+-- | Look up a heap location in an environment.
+lookupHeapLoc :: Name () -> Environment -> HeapLoc
+lookupHeapLoc name (Environment env) =
+    case IntMap.lookup (unUnique $ nameUnique name) env of
+      Nothing  -> throw $ LMachineStringException ("Name " ++ show (nameString name) ++ " missing from environment")
+      Just loc -> loc
+
+emptyHeap :: Heap
+emptyHeap = Heap IntMap.empty 0
+
+-- | Insert a new entry into a heap. The heap grows monotonically with no garbage collection or reuse of heap slots.
+insertInHeap :: Closure -> Heap -> (HeapLoc, Heap)
+insertInHeap cl (Heap h top) =
+    let top' = top + 1
+    in (top', Heap (IntMap.insert top' (Unevaluated cl) h) top')
```

Comment:
I have the impression that we should use dynamic vectors and run the entire thing in `ST`.

Reply:
I mean, later, not in this PR.

Reply:
Good point.  The heap indices always come from some interval [0..top], so some kind of vector would fit that a lot better and would presumably provide constant-time access.  I'll look into that.